### PR TITLE
Add Hive ALTER privilege

### LINKED ISSUE #4253: Presto can not query hive connector (hadoop/hive 3): Unsupported privilege name: ALTER
Tried to let presto-server-333 (334, 336) talk to hive connector (hadoop/hive 3), but all presto queries failed with this error:
`Query 20200625_232316_00000_ev9ef failed: com.google.common.util.concurrent.UncheckedExecutionException: java.lang.IllegalArgumentException: Unsupported privilege name: ALTER`

This is the hadoop/hive version that I am using:
```
HADOOP_VERSION=3.2.1
HIVE_VERSION=3.1.2
```

Full stack trace:
```
com.google.common.util.concurrent.UncheckedExecutionException: com.google.common.util.concurrent.UncheckedExecutionException: java.lang.IllegalArgumentException: Unsupported privilege name: ALTER
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2051)
	at com.google.common.cache.LocalCache.get(LocalCache.java:3951)
	at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3974)
	at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4958)
	at com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4964)
	at io.prestosql.plugin.hive.metastore.cache.CachingHiveMetastore.get(CachingHiveMetastore.java:244)
	at io.prestosql.plugin.hive.metastore.cache.CachingHiveMetastore.listTablePrivileges(CachingHiveMetastore.java:860)
	at io.prestosql.plugin.hive.HiveMetastoreClosure.listTablePrivileges(HiveMetastoreClosure.java:270)
	at io.prestosql.plugin.hive.metastore.SemiTransactionalHiveMetastore.listTablePrivileges(SemiTransactionalHiveMetastore.java:891)
	at io.prestosql.plugin.hive.security.SqlStandardAccessControl.checkTablePermission(SqlStandardAccessControl.java:496)
	at io.prestosql.plugin.hive.security.SqlStandardAccessControl.checkCanSelectFromColumns(SqlStandardAccessControl.java:259)
	at io.prestosql.plugin.hive.security.SystemTableAwareAccessControl.checkCanSelectFromColumns(SystemTableAwareAccessControl.java:90)
	at io.prestosql.plugin.base.classloader.ClassLoaderSafeConnectorAccessControl.checkCanSelectFromColumns(ClassLoaderSafeConnectorAccessControl.java:204)
	at io.prestosql.security.AccessControlManager.lambda$checkCanSelectFromColumns$58(AccessControlManager.java:687)
	at io.prestosql.security.AccessControlManager.catalogAuthorizationCheck(AccessControlManager.java:915)
	at io.prestosql.security.AccessControlManager.checkCanSelectFromColumns(AccessControlManager.java:687)
	at io.prestosql.security.ForwardingAccessControl.checkCanSelectFromColumns(ForwardingAccessControl.java:279)
	at io.prestosql.sql.analyzer.Analyzer.lambda$analyze$0(Analyzer.java:88)
	at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:684)
	at io.prestosql.sql.analyzer.Analyzer.lambda$analyze$1(Analyzer.java:87)
	at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:684)
	at io.prestosql.sql.analyzer.Analyzer.analyze(Analyzer.java:86)
	at io.prestosql.sql.analyzer.Analyzer.analyze(Analyzer.java:75)
	at io.prestosql.execution.SqlQueryExecution.analyze(SqlQueryExecution.java:217)
	at io.prestosql.execution.SqlQueryExecution.<init>(SqlQueryExecution.java:176)
	at io.prestosql.execution.SqlQueryExecution$SqlQueryExecutionFactory.createQueryExecution(SqlQueryExecution.java:705)
	at io.prestosql.dispatcher.LocalDispatchQueryFactory.lambda$createDispatchQuery$0(LocalDispatchQueryFactory.java:121)
	at io.prestosql.$gen.Presto_334____20200626_024818_2.call(Unknown Source)
	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:69)
	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: com.google.common.util.concurrent.UncheckedExecutionException: java.lang.IllegalArgumentException: Unsupported privilege name: ALTER
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2051)
	at com.google.common.cache.LocalCache.get(LocalCache.java:3951)
	at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3974)
	at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4958)
	at com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4964)
	at io.prestosql.plugin.hive.metastore.cache.CachingHiveMetastore.get(CachingHiveMetastore.java:244)
	at io.prestosql.plugin.hive.metastore.cache.CachingHiveMetastore.listTablePrivileges(CachingHiveMetastore.java:860)
	at io.prestosql.plugin.hive.metastore.cache.CachingHiveMetastore.loadTablePrivileges(CachingHiveMetastore.java:912)
	at io.prestosql.plugin.hive.metastore.cache.CachingHiveMetastore.lambda$new$0(CachingHiveMetastore.java:212)
	at com.google.common.cache.CacheLoader$FunctionToCacheLoader.load(CacheLoader.java:165)
	at com.google.common.cache.CacheLoader$1.load(CacheLoader.java:188)
	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3529)
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2278)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2155)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2045)
	... 33 more
Caused by: java.lang.IllegalArgumentException: Unsupported privilege name: ALTER
	at io.prestosql.plugin.hive.metastore.thrift.ThriftMetastoreUtil.parsePrivilege(ThriftMetastoreUtil.java:764)
	at io.prestosql.plugin.hive.metastore.thrift.ThriftHiveMetastore.lambda$listTablePrivileges$62(ThriftHiveMetastore.java:1478)
	at io.prestosql.plugin.hive.metastore.thrift.ThriftMetastoreApiStats.lambda$wrap$0(ThriftMetastoreApiStats.java:42)
	at io.prestosql.plugin.hive.util.RetryDriver.run(RetryDriver.java:130)
	at io.prestosql.plugin.hive.metastore.thrift.ThriftHiveMetastore.listTablePrivileges(ThriftHiveMetastore.java:1455)
	at io.prestosql.plugin.hive.metastore.thrift.BridgingHiveMetastore.listTablePrivileges(BridgingHiveMetastore.java:408)
	at io.prestosql.plugin.hive.metastore.cache.CachingHiveMetastore.loadTablePrivileges(CachingHiveMetastore.java:912)
	at io.prestosql.plugin.hive.metastore.cache.CachingHiveMetastore.lambda$new$0(CachingHiveMetastore.java:212)
	at com.google.common.cache.CacheLoader$FunctionToCacheLoader.load(CacheLoader.java:165)
	at com.google.common.cache.CacheLoader$1.load(CacheLoader.java:188)
	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3529)
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2278)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2155)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2045)
	... 47 more
```

### PR DESCRIPTION
**Pull request by kokosing**
Add Hive ALTER privilege

Fixes #4253

### CODE REVIEW

**Comment by kokosing** [comment]
Fixes https://github.com/prestosql/presto/issues/4253

I don't see ALTER defined here: https://cwiki.apache.org/confluence/display/Hive/SQL+Standard+Based+Hive+Authorization

@jiegzhan where else should look to find what `ALTER` means in Hive 3?

**Comment by findepi** [comment]
shall we have a product test? 
eg grant ALTER in hive and verify reading from information_schema .grants?

**Comment by kokosing** [comment]
> shall we have a product test?

This document defines other privileges that we are missing
https://www.docs4dev.com/docs/en/apache-hive/3.1.1/reference/AuthDev.html#AuthDev-4.1Privilegenames/types:

Also I would like to use something more light weight to test this. Product tests seems to be overkill. Maybe dockerized unit tests like for psql?

**Comment by kokosing** [comment]
https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/sqlstd/SQLAuthorizationUtils.java#L68 defines `"INSERT", "UPDATE", "DELETE", "SELECT"` and this defines more: https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/security/authorization/PrivilegeType.java

**Comment by findepi** [comment]
> Also I would like to use something more light weight to test this. Product tests seems to be overkill. Maybe dockerized unit tests like for psql?

That does not seem much lighter and lacks `env up`.

I suggested product tests since that's the suitable way for testing compatibility issues and #4253 is a compatibility issue.

**Comment by kokosing** [comment]
```
src/main/resources/docker/presto-product-tests/conf/environment/singlenode-kerberos-hdfs-impersonation-with-wire-encryption/hive.properties:hive.security=sql-standard
src/main/resources/docker/presto-product-tests/conf/environment/singlenode-kerberos-hdfs-impersonation/hive.properties:hive.security=sql-standard
src/main/resources/docker/presto-product-tests/conf/environment/singlenode-kerberos-hive-impersonation/hive.properties:hive.security=sql-standard
src/main/resources/docker/presto-product-tests/conf/environment/two-kerberos-hives/hive1.properties:hive.security=sql-standard
src/main/resources/docker/presto-product-tests/conf/environment/two-kerberos-hives/hive2.properties:hive.security=sql-standard
src/main/resources/docker/presto-product-tests/conf/environment/two-mixed-hives/hive1.properties:hive.security=sql-standard
```

There is no hive3 setup with sql-standard enabled.

@jiegzhan would you be able to test this pull request in your environment and see if there is anything else that we are missing.

**Comment by findepi** [comment]
> There is no hive3 setup with sql-standard enabled.


@kokosing 

These

```
src/main/resources/docker/presto-product-tests/conf/environment/singlenode-kerberos-hdfs-impersonation-with-wire-encryption/hive.properties:hive.security=sql-standard
src/main/resources/docker/presto-product-tests/conf/environment/singlenode-kerberos-hdfs-impersonation/hive.properties:hive.security=sql-standard
src/main/resources/docker/presto-product-tests/conf/environment/singlenode-kerberos-hive-impersonation/hive.properties:hive.security=sql-standard
```

are run with hdp2.6, hdp3.1 (hive 3) and cdh5

**Comment by jiegzhan** [comment]
Settings:
```
[root@amazon-linux-2-jie-dev presto]# echo $JAVA_HOME
/usr/lib/jvm/jre-11-openjdk
[root@amazon-linux-2-jie-dev presto]# echo $MAVEN_HOME
/usr/lib/apache-maven-3.6.3
[root@amazon-linux-2-jie-dev presto]#
```

Clone your repo and checkout 003_alter branch:
```
git clone https://github.com/kokosing/presto.git
git checkout -b 003_alter origin/origin/master/003_alter
mvn clean install -DskipTests
```

Build failed with this error message:
```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  8.056 s
[INFO] Finished at: 2020-06-29T20:26:49Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:compile (default-compile) on project presto-spi: Fatal error compiling: error: release version 11 not supported -> [Help 1]
```

Any idea how to address this? Thanks.

**Comment by findepi** [comment]
For the privileges that have no meaning in Presto, is there a point in defining them?
User will be able to GRANT them (good), but they will expect them to take effect and they can be surprised they do not (bad).

Since we effectively ignore them, we could ignore them early, when mapping from thrift responses to Presto permissions model.

**Comment by jiegzhan** [comment]
@kokosing @electrum I see this PR got approved but still open, is it possible to have this in next release in order to make presto work with hive-3-metastore? Thanks.

**Comment by kokosing** [comment]
@jiegzhan Currently the product tests is failing because `GRANT ALTER` in Hive errors with `FAILED: Execution Error, return code 1 from org.apache.hadoop.hive.ql.exec.DDLTask. Unsupported privilege type ALTER`. Can you please tell me how I can grant ALTER privilege to user?

What Hadoop distribution and deployments are you using? Are you using any other tools that could affect authorization policies (like Impala)?

**Comment by jiegzhan** [comment]
@kokosing, I setup hadoop-3.2.1 and hive-3.1.2, then connect presto to it. That's all the tools that I am using.

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestSqlStandardAccessControlChecks.java [hunk 1]
```diff
@@ -240,4 +240,13 @@ public void testAccessControlFilterColumns()
         aliceExecutor.executeQuery(format("GRANT SELECT ON %s TO bob", tableName));
         assertThat(bobExecutor.executeQuery(format("SELECT * FROM information_schema.columns WHERE table_name = '%s'", tableName))).hasRowsCount(2);
     }
+
+    @Test(groups = {AUTHORIZATION, PROFILE_SPECIFIC_TESTS})
+    public void testHiveAlterPrivilege()
+    {
+        onHive().executeQuery(format("GRANT ALTER ON TABLE %s TO USER bob", tableName));
```
Comment (jiegzhan) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestSqlStandardAccessControlChecks.java:247]: I think we don't need these two words: TABLE, USER
examples here: https://prestosql.io/docs/333/sql/grant.html#examples
`GRANT INSERT, SELECT ON orders TO alice;`

**Pull request closed without merging**